src/names/scopes.md

```diff
@@ -1,3 +1,329 @@
+
+Item names from outer modules are not in scope within a nested module.
+A [path] may be used to refer to an item in another module.
+
+### Associated item scopes
+
+[Associated items] are not scoped and can only be referred to by using a [path] leading from the type they are associated with.
+[Methods] can also be referred to via [call expressions].
+
+Similar to items within a module or block, it is an error to introduce an item within a trait or implementation that is a duplicate of another item in the trait or impl in the same namespace.
+
+## Pattern binding scopes
+
+The scope of a local variable [pattern] binding depends on where it is used:
+
+* [`let` statement] bindings range from just after the `let` statement until the end of the block where it is declared.
+* [Function parameter] bindings are within the body of the function.
+* [Closure parameter] bindings are within the closure body.
+* [`for`] and [`while let`] bindings are within the loop body.
+* [`if let`] bindings are within the consequent block.
+* [`match` arms] bindings are within the [match guard] and the match arm expression.
+
+Local variable scopes do not extend into item declarations.
+<!-- Not entirely, see https://github.com/rust-lang/rust/issues/33118 -->
+
+### Pattern binding shadowing
+
+Pattern bindings are allowed to shadow any name in scope with the following exceptions which are an error:
+
+* [Const generic parameters]
+* [Static items]
+* [Const items]
+* Constructors for [structs] and [enums]
+
+The following example illustrates how local bindings can shadow item declarations:
+
+```rust
+fn shadow_example() {
+    // Since there are no local variables in scope yet, this resolves to the function.
+    foo(); // prints `function`
+    let foo = || println!("closure");
+    fn foo() { println!("function"); }
+    // This resolves to the local closure since it shadows the item.
+    foo(); // prints `closure`
+}
+```
+
+## Generic parameter scopes
+
+Generic parameters are declared in a [_GenericParams_] list.
+The scope of a generic parameter is within the item it is declared on.
+
+All parameters are in scope within the generic parameter list regardless of the order they are declared.
+The following shows some examples where a parameter may be referenced before it is declared:
+
+```rust
+// The 'b bound is referenced before it is declared.
+fn params_scope<'a: 'b, 'b>() {}
+
+# trait SomeTrait<const Z: usize> {}
+// The const N is referenced in the trait bound before it is declared.
+fn f<T: SomeTrait<N>, const N: usize>() {}
+```
+
+Generic parameters are also in scope for type bounds and where clauses, for example:
+
+```rust
+# trait SomeTrait<'a, T> {}
+// The <'a, U> for `SomeTrait` refer to the 'a and U parameters of `bounds_scope`.
+fn bounds_scope<'a, T: SomeTrait<'a, U>, U>() {}
+
+fn where_scope<'a, T, U>()
+    where T: SomeTrait<'a, U>
+{}
+```
+
+Generic scopes do not extend into [items] declared inside a function.
```

Comment:
This is kind of a philosophical question, but they do extend (*), it's just an error to use a parameter from an outer item.

(*) until a `mod` item is encountered, which is a true barrier for names.

Reply:
That seems to be an implementation choice, though?  Like, the choice is:

1) Generic names do not extend into an inner item. Using a parent name will be a "name not found" error.
2) Generic names do extend into an inner item, but can be shadowed. Using a parent name without shadowing it will be a "can't use generic parameter from outer function" error.

I felt like the first approach was a little easier to explain, even if that's not how `rustc` works (which I see as just a way to provide better errors, or maybe just easier to implement).

That being said, I reworded it to say it is an error.